talk/SessionLoader.swift

```diff
@@ -18,27 +18,51 @@ import Foundation
 /// 3. Decodable 的型別
 ///
 
+enum LoaderError {
+    case urlSessionError(Error)
+    case noData
+    case decodeError(Error)
+    
+    var message: String {
+        switch self {
+        case let .urlSessionError(err):
+            return err.localizedDescription
+        case .noData:
+            return "Server return empty data"
+        case let .decodeError(err):
+            return err.localizedDescription
+        }
+    }
+}
+
 class SessionLoader<DecodeType : Decodable> {
+    
    init(request: URLRequest) {
         self.request = request
     }
     
-    
     let request: URLRequest
     
     @discardableResult
-    func loadData(success: ((DecodeType) -> Void)?) -> URLSessionTask {
+    func loadData(
+        success: ((DecodeType) -> Void)?,
+        failure: ((LoaderError) -> Void)?
+    ) -> URLSessionTask {
     
-        let task = URLSession.shared.dataTask(with: self.request) { data, res, err in
+        let task = URLSession.shared.dataTask(with: request) { data, res, err in
             if err != nil {
+                failure?(.urlSessionError(err!))
+                return
+            }
```

Comment:
```suggestion
            if let err = err {
                failure?(.urlSessionError(err))
                return
            }
```
使用 Swift 語言的 `if let` 語法簡化不必要的 `err!`，可以提升閱讀性

Reply:
都可
這兩個寫法在我個人閱讀而言 是等意的